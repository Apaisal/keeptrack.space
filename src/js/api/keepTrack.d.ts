/* eslint-disable no-unused-vars */
export declare interface SatObject {
  id: number;
  satId?: number;
  position: { x: number; y: number; z: number };
  static?: boolean;
  missile?: boolean;
  type?: string;
  name?: string;
  intlDes?: string;
  TLE1: string;
  TLE2: string;
  SCC_NUM: string;
  active: boolean;
  C?: string;
  LS?: string;
  LV?: string;
  ON: string;
  OT: number;
  TLE1: string;
  TLE2: string;
  R: string;
  URL?: string;
  O?: string;
  U?: string;
  P?: string;
  LM?: string;
  DM?: string;
  Pw?: string;
  Li?: string | number;
  Con?: string;
  M?: string;
  S1?: string;
  S2?: string;
  S3?: string;
  S4?: string;
  S5?: string;
  S6?: string;
  S7?: string;
  inclination: number;
  lon: number;
  perigee: number;
  apogee: number;
  period: number;
  meanMotion: number;
  semimajorAxis: number;
  eccentricity: number;
  raan: number;
  argPe: number;
  inView: number;
  velocity: {
    total: number;
    x: number;
    y: number;
    z: number;
  };
  getTEARR?: any;
  getAltitude?: any;
  getDirection?: any;
  vmag?: number;
  associates?: any;
  maneuver?: any;
  constellation?: any;
  ORPO?: any;
  FMISSED?: any;
  NOTES?: any;
  TTP?: any;
  isInSun?: any;
  inSun?: any;
  desc?: string;
  marker?: boolean;
  isRadarData?: boolean;
  isInGroup?: boolean;
  missileComplex?: number;
}

export interface SensorObject {
  obsmaxel2?: any;
  obsmaxrange2?: any;
  obsminrange2?: any;
  obsminel2?: any;
  obsmaxaz2?: any;
  obsminaz2?: any;
  alt: number;
  beamwidth?: number;
  changeObjectInterval?: number;
  country: string;
  lat: number;
  linkAehf?: boolean;
  linkWgs?: boolean;
  lon: number;
  name: string;
  alt: number;
  observerGd: {
    lat: number;
    lon: number;
    alt: number;
  };
  obsmaxaz: number;
  obsmaxel: number;
  obsmaxrange: number;
  obsminaz: number;
  obsminel: number;
  obsminrange: number;
  shortName: string;
  staticNum: number;
  sun: string;
  type?: string;
  url?: string;
  volume: boolean;
  zoom: string;
}

export interface Lla {
  lat: number;
  lon: number;
  alt: number;
}

export interface Rae {
  rng: number;
  az: number;
  el: number;
}

export interface InView {
  inView: boolean;
}

export interface EarthObject {
  init: any;
  draw: any;
  program: any;
  update: any;
  drawOcclusion: any;
  earthJ: number;
  earthEra: number;
  lightDirection: any;
  pos: [number, number, number];
  loaded: boolean;
  sunvar: any;
  shader: any;
  specularMap: any;
  imgHiRes: any;
  isHiResReady: boolean;
  nightImgHiRes: any;
  nightImg: any;
  bumpMap: any;
  loadHiResNight: any;
  loadHiRes: any;
  isUseHiRes: boolean;
  lightDirection: [number, number, number];
}

export interface keepTrackApiInterface {
  html: (strings: TemplateStringsArray, ...placeholders: any[]) => string;
  register: (params: { method: string; cbName: string; cb: any }) => void;
  unregister: (params: { method: string; cbName: string }) => void;
  callbacks: any;
  methods: any;
  programs: any;
}
