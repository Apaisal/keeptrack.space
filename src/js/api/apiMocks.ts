import { satObject, sensorObject } from './keepTrack';

export const defaultSat: satObject = {
  active: true,
  SCC_NUM: '00005',
  C: 'ISS',
  LS: 'TTMTR',
  LV: 'Proton-K',
  ON: 'ISS (ZARYA)',
  OT: 1,
  TLE1: '1 25544U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991',
  TLE2: '2 25544  51.6423 168.5744 0001475 184.3976 313.3642 15.48839820294053',
  R: '99.0524',
  URL: 'http://www.esa.int/export/esaHS/ESAI2X0VMOC_iss_0.html',
  O: 'National Aeronautics and Space Administration (NASA)/Multinational',
  U: 'Government',
  P: 'Space Science',
  LM: '',
  DM: '',
  Pw: '',
  Li: 30,
  Con: 'Boeing Satellite Systems (prime)/Multinational',
  M: 'Final size of a Boeing 747; first component.',
  S1: 'http://www.esa.int/export/esaHS/ESAI2X0VMOC_iss_0.html',
  S2: '',
  S3: 'http://www.boeing.com/defense-space/space/spacestation/overview/',
  S4: '',
  S5: '',
  S6: '',
  S7: '',
  inclination: 51.6423,
  longitude: 168.5744,
  perigee: 1475,
  apogee: 184.3976,
  period: 313.3642,
  meanMotion: 15.48839820294053,
  semimajorAxis: 21203.40407588,
  eccentricity: 0.00003453,
  raan: 0,
  argPe: 0,
  inview: 1,
  velocity: {
    total: 0,
    x: 0,
    y: 0,
    z: 0,
  },
  getTEARR: () => ({
    lon: 168.5744,
    lat: 51.6423,
    alt: 1475,
  }),
  getAltitude: () => 0,
  getDirection: () => 'N',
};

export const defaultSensor: sensorObject = {
  alt: 0.060966,
  beamwidth: 2,
  changeObjectInterval: 1000,
  country: 'United States',
  lat: 41.754785,
  linkAehf: true,
  linkWgs: true,
  lon: -70.539151,
  name: 'Cape Cod AFS, Massachusetts',
  observerGd: { lat: 0.7287584767123405, lon: -1.2311404365114507, alt: 0.060966 },
  obsmaxaz: 227,
  obsmaxel: 85,
  obsmaxrange: 5556,
  obsminaz: 347,
  obsminel: 3,
  obsminrange: 200,
  shortName: 'COD',
  staticNum: 0,
  sun: 'No Impact',
  type: 'Phased Array Radar',
  url: 'http://www.radartutorial.eu/19.kartei/01.oth/karte004.en.html',
  volume: false,
  zoom: 'leo',
};

export const useMockWorkers = () => {
  class Worker {
    url: any;
    onmessage: (msg: any) => void;
    constructor(stringUrl: any) {
      this.url = stringUrl;
      this.onmessage = () => {
        return {
          data: {}
        }
      };
    }

    postMessage(msg: any) {
      if (msg.dat) {
        this.onmessage({          
          data: {
            extraData: JSON.stringify([defaultSat, defaultSensor]),
            satPos: [0,0,0,0,0,0],
            satVel: [0,0,0,0,0,0],
          }
        });
      }
      this.onmessage({
        data: {
          satPos: [0,0,0,0,0,0],
          satVel: [0,0,0,0,0,0],
        }
      });
    }
  }

  (<any>window).Worker = Worker;
}

export const keepTrackApiStubs = {
  programs: {
    adviceManager: {
      adviceList: {
        satViewDisabled: jest.fn(),
        editSatDisabled: jest.fn(),
        survFenceDisabled: jest.fn(),
        bubbleDisabled: jest.fn(),
        satFOVDisabled: jest.fn(),
        planetariumDisabled: jest.fn(),
        mapDisabled: jest.fn(),
      },
      adviceCount: {
        socrates: 0,
      },
      showAdvice: jest.fn(),
      adviceArray: (<any[]>[]),
    },
    mainCamera: {
      cameraType: {
        current: 1,
        satellite: 1,
        fixedToSat: 2,
      },
      camSnap: jest.fn(),
      changeZoom: jest.fn(),
      latToPitch: jest.fn(),
      longToYaw: jest.fn(),
      autoRotate: jest.fn(),
    },
    ColorScheme: {
      reloadColors: jest.fn(),      
    },
    drawManager: {
      glInit: jest.fn(),
      gl: global.mocks.glMock,
      selectSatManager: {
        selectSat: jest.fn(),
      },
    },    
    dotsManager: {
      starIndex1: 0,
      pickReadPixelBuffer: [
        0,
        0,
        0,
      ],
    },    
    groups: null,
    groupsManager: {
      Canada: null,
      Japan: null,
      France: null,
      China: null,
      India: null,
      Israel: null,
      Russia: null,
      UnitedKingdom: null,
      UnitedStates: null,
      SpaceStations: {
        sats: [
          {satId: 1, SCC_NUM: '25544'},
          {satId: 1, SCC_NUM: '25544'},
        ],
        updateOrbits: jest.fn(),
      },
      clearSelect: jest.fn(),
      createGroup: () => ({
        updateOrbits: jest.fn(),
      }),
      selectGroup: jest.fn(),
    },    
    lineManager: {
      create: jest.fn(),
      removeStars: jest.fn(),
      getLineListLen: jest.fn(),
    },
    missileManager: {
      getMissileTEARR: jest.fn(),
    },
    objectManager: {
      isLaunchSiteManagerLoaded: true,
      launchSiteManager: {
        launchSiteList: [
          'AFWTR'
        ],
      },
      selectedSat: -1,
      setSelectedSat: jest.fn(),
      staticSet: {
        id: 0,
      },
      analSatSet: {
        id: 0,
      },
      radarDataSet: {
        id: 0,
      },
      missileSet: {
        id: 0,
      },
      fieldOfViewSet: {
        id: 0,
      },
    },
    orbitManager: {
      updateOrbitBuffer: jest.fn(),
      clearInViewOrbit: jest.fn(),
      removeInViewOrbit: jest.fn(),
      historyOfSatellitesPlay: jest.fn(),
    },
    satellite: {
      currentTEARR: {
        az: 0,
        el: 0,
        rng: 0,
      },
      sgp4: () => ([0,0,0]),
      setobs: jest.fn(),
      setTEARR: jest.fn(),
      altitudeCheck: jest.fn(),
      degreesLong: (num: number) => num,
      degreesLat: (num: number) => num,
      currentEpoch: () => ['21', '150'],
      getOrbitByLatLon: () => ['1 25544U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991', '2 25544  51.6423 168.5744 0001475 184.3976 313.3642 15.48839820294053'],
      nextpassList: () => [new Date().getTime(), new Date().getTime() + 1],
      twoline2satrec: () => ({ jdsatepoch: 2458000.5 }),
      getRae: () => [0, 0, 0],
      checkIsInFOV: () => true,
      findCloseObjects: () => '',
      findBestPasses: () => jest.fn(),
      nextpass: () => '2021-09-21 20:00:00Z',
      getDOPsTable: jest.fn(),
    },
    satSet: {
      satCruncher: {
        postMessage: jest.fn(),
      },
      satData: [
        defaultSat,
        defaultSat,
      ],
      getIdFromObjNum: jest.fn(),
      getSensorFromSensorName: () => 0,
      getSatExtraOnly: () => defaultSat,
      getSatFromObjNum : () => defaultSat,
      getSat: () => defaultSat,
      initGsData: jest.fn(),
      setColorScheme: jest.fn(),
      getSatInViewOnly: () => defaultSat,
      default: '',
      insertNewAnalystSatellite: jest.fn(),
    },
    searchBox: {
      hideResults: jest.fn(),
      fillResultBox: jest.fn(),
    },
    sensorFov: {
      enableFovView: jest.fn(),
    },
    sensorManager: {
      checkSensorSelected: () => false,
      currentSensor: {
        alt: 0,
        lat: 0,
        lon: 0,
        shortName: 'COD',
      },
    },
    starManager: {
      init: jest.fn(),
    },
    timeManager: {
      propOffset: 0,
      propRate: 0,
      selectedDate: 0,
      propTime: () => 0,
      getDayOfYear: () => 0,
    },
    uiManager: {
      hideSideMenus: jest.fn(),
      toast: jest.fn(),
      colorSchemeChangeAlert: jest.fn(),
      bottomIconPress: jest.fn(),
      legendColorsChange: jest.fn(),
      doSearch: jest.fn(),
      loadStr: jest.fn(),
      searchToggle: jest.fn(),
      legendMenuChange: jest.fn(),
      clearRMBSubMenu: jest.fn(),
      updateURL: jest.fn(),
    },
    watchlist: {
      updateWatchlist: jest.fn(),
    },
  },
};

keepTrackApiStubs.programs.groupsManager.Canada = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Japan = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.France = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.China = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.India = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Israel = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.Russia = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.UnitedKingdom = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groupsManager.UnitedStates = keepTrackApiStubs.programs.groupsManager.SpaceStations;
keepTrackApiStubs.programs.groups = keepTrackApiStubs.programs.groupsManager;