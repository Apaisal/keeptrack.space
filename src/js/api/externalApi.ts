interface keepTrackApiInterface {
  register: (params: { method: string; cbName: string; cb: () => void }) => void;
  unregister: (params: { method: string; cbName: string }) => void;
  callbacks: any;
  methods: any;
  programs: any;
}

const keepTrackApi: keepTrackApiInterface = {
  register: (params: { method: string; cbName: string; cb: () => void }) => {
    // If this is a valid callback
    if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
      // Add the callback
      keepTrackApi.callbacks[params.method].push({ name: params.cbName, cb: params.cb });
    } else {
      throw new Error(`Invalid callback "${params.method}"!`);
    }
    return;
  },
  unregister: (params: { method: string; cbName: string }) => {
    // If this is a valid callback
    if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
      for (let i = 0; i < keepTrackApi.callbacks[params.method].length; i++) {
        if (keepTrackApi.callbacks[params.method][i].name == params.cbName) {
          keepTrackApi.callbacks[params.method].splice(i, 1);
          return;
        }
      }
      // If we got this far, it means we couldn't find the callback
      throw new Error(`Callback "${params.cbName} not found"!`);
    } else {
      // Couldn't find the method
      throw new Error(`Invalid callback "${params.method}"!`);
    }
  },
  callbacks: {
    selectSatData: [],
    updateSelectBox: [],
  },
  methods: {
    selectSatData: (sat: any, satId: number) => {
      for (let i = 0; i < keepTrackApi.callbacks.selectSatData.length; i++) {
        keepTrackApi.callbacks.selectSatData[i].cb(sat, satId);
      }
    },
    updateSelectBox: (sat: any) => {
      for (let i = 0; i < keepTrackApi.callbacks.updateSelectBox.length; i++) {
        keepTrackApi.callbacks.updateSelectBox[i].cb(sat);
      }
    },
  },
  programs: {
    timeManager: {},
    settingsManager: {},
    ColorScheme: {},
    drawManager: {},
    missileManager: {},
    objectManager: {},
    orbitManager: {},
    photoManager: {},
    satSet: {},
    satellite: {},
    searchBox: {},
    sensorManager: {},
    starManager: {},
    sMM: {},
    uiManager: {},
    uiInput: {},
  },
};

(<any>window).keepTrackApi = keepTrackApi;
export { keepTrackApi };
