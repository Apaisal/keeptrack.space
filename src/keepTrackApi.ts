/* eslint-disable class-methods-use-this */
import { BaseObject, DetailedSatellite, DetailedSensor, Milliseconds } from 'ootk';
import { keepTrackContainer } from './container';
import { Constructor, GroupsManager, KeepTrackApiEvents, OrbitManager, SensorManager, Singletons, UiManager } from './interfaces';
import { KeepTrackPlugin } from './plugins/KeepTrackPlugin';
import { SoundNames } from './plugins/sounds/SoundNames';
import { SoundManager } from './plugins/sounds/sound-manager';
import { SettingsManager } from './settings/settings';
import { Camera } from './singletons/camera';
import { CatalogManager } from './singletons/catalog-manager';
import { MissileObject } from './singletons/catalog-manager/MissileObject';
import { ColorSchemeManager } from './singletons/color-scheme-manager';
import { DotsManager } from './singletons/dots-manager';
import { LineManager } from './singletons/draw-manager/line-manager';
import { MeshManager } from './singletons/draw-manager/mesh-manager';
import { errorManagerInstance } from './singletons/errorManager';
import { HoverManager } from './singletons/hover-manager';
import { InputManager } from './singletons/input-manager';
import { PanTouchEvent, TapTouchEvent } from './singletons/input-manager/touch-input';
import { Scene } from './singletons/scene';
import { StarManager } from './singletons/starManager';
import { TimeManager } from './singletons/time-manager';
import { WebGLRenderer } from './singletons/webgl-renderer';
import { SatMath } from './static/sat-math';
import { SensorMath } from './static/sensor-math';

declare global {
  interface Window {
    settingsManager: SettingsManager;
    jQuery: unknown;
    $: unknown;
    gremlins: any;
    randomizer: any;
    keepTrackApi: KeepTrackApi;
    dataLayer: any; // For Google Tag Manager
    _numeric: any;
    satellite: SatMath;
    M: any;
  }
}

type KeepTrackApiEventArguments = {
  [KeepTrackApiEvents.bottomMenuClick]: [string];
  [KeepTrackApiEvents.hideSideMenus]: [];
  [KeepTrackApiEvents.nightToggle]: [WebGL2RenderingContext, WebGLTexture, WebGLTexture];
  [KeepTrackApiEvents.orbitManagerInit]: [];
  [KeepTrackApiEvents.drawManagerLoadScene]: [];
  [KeepTrackApiEvents.drawOptionalScenery]: [];
  [KeepTrackApiEvents.updateLoop]: [];
  [KeepTrackApiEvents.rmbMenuActions]: [string, number];
  [KeepTrackApiEvents.rightBtnMenuAdd]: [];
  [KeepTrackApiEvents.updateDateTime]: [Date];
  [KeepTrackApiEvents.uiManagerFinal]: [];
  [KeepTrackApiEvents.resetSensor]: [];
  [KeepTrackApiEvents.setSensor]: [DetailedSensor | string, number];
  [KeepTrackApiEvents.changeSensorMarkers]: [string];
  [KeepTrackApiEvents.altCanvasResize]: [];
  [KeepTrackApiEvents.endOfDraw]: [Milliseconds];
  [KeepTrackApiEvents.onWatchlistUpdated]: [number[]];
  [KeepTrackApiEvents.staticOffsetChange]: [number];
  [KeepTrackApiEvents.onLineAdded]: [LineManager];
  [KeepTrackApiEvents.sensorDotSelected]: [DetailedSensor];
  [KeepTrackApiEvents.canvasMouseDown]: [MouseEvent];
  [KeepTrackApiEvents.touchStart]: [TapTouchEvent | PanTouchEvent];
  [KeepTrackApiEvents.onCruncherMessage]: [];
  [KeepTrackApiEvents.onCruncherReady]: [];
  [KeepTrackApiEvents.onHelpMenuClick]: [];
  [KeepTrackApiEvents.onKeepTrackReady]: [];
  [KeepTrackApiEvents.selectSatData]: [DetailedSatellite | MissileObject | BaseObject, number];
  [KeepTrackApiEvents.setSecondarySat]: [DetailedSatellite, number];
  [KeepTrackApiEvents.uiManagerInit]: [];
  [KeepTrackApiEvents.uiManagerOnReady]: [];
  [KeepTrackApiEvents.updateSelectBox]: [DetailedSatellite | MissileObject];
};

interface KeepTrackApiRegisterParams<T extends KeepTrackApiEvents> {
  event: T;
  cbName: string;
  cb: (...args: KeepTrackApiEventArguments[T]) => void;
}

type rmbMenuItem = {
  /**
   * Element ID of the main menu item
   */
  elementIdL1: string;
  /**
   * Element ID of the sub menu container
   */
  elementIdL2: string;
  /**
   * The sorting order for the menus
   */
  order: number;
  /**
   * Determines if the menu item is visible when right clicking on the earth
   */
  isRmbOnEarth: boolean;
  /**
   * Determines if the menu item is visible when right clicking off the earth
   */
  isRmbOffEarth: boolean;
  /**
   * Determines if the menu item is visible when right clicking on a satellite
   */
  isRmbOnSat: boolean;
};

export class KeepTrackApi {
  /**
   * Unregisters all events in the KeepTrackApi. Used for testing.
   */
  unregisterAllEvents() {
    for (const event of Object.values(KeepTrackApiEvents)) {
      this.events[event] = [];
    }
  }

  containerRoot = <HTMLDivElement>null;
  isInitialized = false;
  loadedPlugins = <KeepTrackPlugin[]>[];
  rmbMenuItems = <rmbMenuItem[]>[];
  events = <any>{
    altCanvasResize: [],
    nightToggle: [],
  };

  methods = {
    nightToggle: (gl: WebGL2RenderingContext, nightTexture: WebGLTexture, texture: WebGLTexture) => {
      this.events.nightToggle.forEach((cb: any) => cb.cb(gl, nightTexture, texture));
    },
    altCanvasResize: (): boolean => this.events.altCanvasResize.some((cb: any) => cb.cb()),
  };

  runEvent<T extends KeepTrackApiEvents>(event: T, ...args: KeepTrackApiEventArguments[T]) {
    this.verifyEvent_(event);

    if (event === KeepTrackApiEvents.bottomMenuClick) {
      keepTrackApi.getSoundManager().play(SoundNames.BEEP);
    }

    (<KeepTrackApiRegisterParams<T>[]>this.events[event]).forEach((cb: KeepTrackApiRegisterParams<T>) => cb.cb(...args));
  }

  /** If the callback does not exist, create it */
  private verifyEvent_(event: KeepTrackApiEvents) {
    if (typeof this.events[event] === 'undefined') {
      this.events[event] = [];
    }
  }

  getPlugin<T extends KeepTrackPlugin>(pluginClass: Constructor<T>): T | null {
    if (this.loadedPlugins.some((plugin: KeepTrackPlugin) => plugin instanceof pluginClass)) {
      return this.loadedPlugins.find((plugin: KeepTrackPlugin) => plugin instanceof pluginClass) as T;
    }
    return null;
  }

  /** This is not a standard function. It is used in development for formatting template literals.
   * example: keepTrackApi.html\`\<div>example\</div>\`
   * TODO: This should be a static method
   */
  // eslint-disable-next-line class-methods-use-this
  html(strings: TemplateStringsArray, ...placeholders: any[]) {
    for (const placeholder of placeholders) {
      if (typeof placeholder !== 'string') {
        errorManagerInstance.error(new Error('Invalid input'), 'keepTrackApi.html');
      }
    }
    return String.raw(strings, ...placeholders);
  }

  /** This is not a standard function. It is used in development for formatting template literals.
   * example: keepTrackApi.glsl\`uniform float example\`
   * TODO: This should be a static method
   */
  // eslint-disable-next-line class-methods-use-this
  glsl(literals: TemplateStringsArray, ...placeholders: Array<any>): string {
    let str = '';
    for (let i = 0; i < placeholders.length; i++) {
      str += literals[i];
      str += placeholders[i];
    }
    str += literals[literals.length - 1];
    return str;
  }

  /**
   * Registers a callback function for a specific event.
   * @param {KeepTrackApiEvents} params.event - The name of the event to register the callback for.
   * @param {string} params.cbName - The name of the callback function.
   * @param params.cb - The callback function to register.
   * @throws An error if the event is invalid.
   */
  register<T extends KeepTrackApiEvents>(params: { event: T; cbName: string; cb: (...args: KeepTrackApiEventArguments[T]) => void }) {
    this.verifyEvent_(params.event);

    // Add the callback
    this.events[params.event].push({ name: params.cbName, cb: params.cb });
  }

  unregister(params: { event: KeepTrackApiEvents; cbName: string }) {
    for (let i = 0; i < this.events[params.event].length; i++) {
      if (this.events[params.event][i].name == params.cbName) {
        this.events[params.event].splice(i, 1);
        return;
      }
    }
    // If we got this far, it means we couldn't find the callback
    errorManagerInstance.error(new Error(`Callback "${params.cbName} not found"!`), 'keepTrackApi.unregister');
  }

  getSoundManager = () => keepTrackContainer.get<SoundManager>(Singletons.SoundManager);
  getStarManager = () => keepTrackContainer.get<StarManager>(Singletons.StarManager);
  getRenderer = () => keepTrackContainer.get<WebGLRenderer>(Singletons.WebGLRenderer);
  getScene = () => keepTrackContainer.get<Scene>(Singletons.Scene);
  getCatalogManager = () => keepTrackContainer.get<CatalogManager>(Singletons.CatalogManager);
  getSensorManager = () => keepTrackContainer.get<SensorManager>(Singletons.SensorManager);
  getUiManager = () => keepTrackContainer.get<UiManager>(Singletons.UiManager);
  getInputManager = () => keepTrackContainer.get<InputManager>(Singletons.InputManager);
  getGroupsManager = () => keepTrackContainer.get<GroupsManager>(Singletons.GroupsManager);
  getTimeManager = () => keepTrackContainer.get<TimeManager>(Singletons.TimeManager);
  getOrbitManager = () => keepTrackContainer.get<OrbitManager>(Singletons.OrbitManager);
  getColorSchemeManager = () => keepTrackContainer.get<ColorSchemeManager>(Singletons.ColorSchemeManager);
  getDotsManager = () => keepTrackContainer.get<DotsManager>(Singletons.DotsManager);
  getSensorMath = () => keepTrackContainer.get<SensorMath>(Singletons.SensorMath);
  getLineManager = () => keepTrackContainer.get<LineManager>(Singletons.LineManager);
  getHoverManager = () => keepTrackContainer.get<HoverManager>(Singletons.HoverManager);
  getMainCamera = () => keepTrackContainer.get<Camera>(Singletons.MainCamera);
  getMeshManager = () => keepTrackContainer.get<MeshManager>(Singletons.MeshManager);
}

export const keepTrackApi = new KeepTrackApi();
