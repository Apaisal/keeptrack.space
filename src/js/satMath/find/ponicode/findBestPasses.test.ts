import { keepTrackApiStubs } from '@app/js/api/apiMocks';
import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { KeepTrackPrograms } from '@app/js/api/keepTrackTypes';
import * as findBestPasses from '@app/js/satMath/find/findBestPasses';
import * as SpaceObjectType from '../../../api/SpaceObjectType';

keepTrackApi.programs = <KeepTrackPrograms>(<unknown>{ ...keepTrackApi.programs, ...keepTrackApiStubs.programs });

// @ponicode
describe('findBestPasses.findBestPasses', () => {
  test('0', () => {
    let result: any = findBestPasses.findBestPasses('Jean-Philippe,,George', {
      static: true,
      linkIridium: true,
      linkGalileo: false,
      linkStarlink: true,
      obsmaxel2: 1,
      obsmaxrange2: 550,
      obsminrange2: 550,
      obsminel2: 380,
      obsmaxaz2: -100,
      obsminaz2: 100,
      alt: 3,
      beamwidth: 800,
      changeObjectInterval: 1,
      country: 'FR',
      lat: 90,
      linkAehf: true,
      linkWgs: false,
      lon: 45,
      name: 'Jean-Philippe',
      observerGd: { lat: -45, lon: 180, alt: 1 },
      obsmaxaz: 100,
      obsmaxel: 1,
      obsmaxrange: 100,
      obsminaz: -5.48,
      obsminel: 10.0,
      obsminrange: 0,
      shortName: 'http://www.example.com/route/123?foo=bar',
      staticNum: 1,
      sun: '4.0.0-beta1\t',
      type: SpaceObjectType.SpaceObjectType.LAUNCH_VEHICLE_MANUFACTURER,
      url: 'https://twitter.com/path?abc',
      volume: true,
      zoom: 'leo',
    });
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    let result: any = findBestPasses.findBestPasses(',', {
      static: true,
      linkIridium: false,
      linkGalileo: false,
      linkStarlink: false,
      obsmaxel2: -5.48,
      obsmaxrange2: 4,
      obsminrange2: 380,
      obsminel2: 50,
      obsmaxaz2: 100,
      obsminaz2: 1,
      alt: 50,
      beamwidth: 10,
      changeObjectInterval: -5.48,
      country: 'United States',
      lat: 0,
      linkAehf: true,
      linkWgs: true,
      lon: -180,
      name: 'Jean-Philippe',
      observerGd: { lat: -45, lon: -90, alt: 1 },
      obsmaxaz: 100,
      obsmaxel: 100,
      obsmaxrange: 100,
      obsminaz: 0,
      obsminel: 0.5,
      obsminrange: -100,
      shortName: 'http://base.com',
      staticNum: 10,
      sun: '1.0.0',
      type: SpaceObjectType.SpaceObjectType.PAYLOAD,
      url: 'https://',
      volume: false,
      zoom: 'geo',
    });
    expect(result).toMatchSnapshot();
  });

  test('2', () => {
    let result: any = findBestPasses.findBestPasses('Jean-Philippe,George', {
      static: false,
      linkIridium: false,
      linkGalileo: false,
      linkStarlink: true,
      obsmaxel2: -5.48,
      obsmaxrange2: 400,
      obsminrange2: 520,
      obsminel2: 520,
      obsmaxaz2: 1,
      obsminaz2: -5.48,
      alt: 50,
      beamwidth: 16,
      changeObjectInterval: -100,
      country: 'United States',
      lat: 90,
      linkAehf: false,
      linkWgs: false,
      lon: 0,
      name: 'Michael',
      observerGd: { lat: -180, lon: -45, alt: 1 },
      obsmaxaz: 0,
      obsmaxel: 100,
      obsmaxrange: 380,
      obsminaz: 100,
      obsminel: -0.5,
      obsminrange: 1,
      shortName: 'ponicode.com',
      staticNum: 10,
      sun: 'v1.2.4',
      type: SpaceObjectType.SpaceObjectType.MECHANICAL,
      url: 'http://base.com',
      volume: true,
      zoom: 'leo',
    });
    expect(result).toMatchSnapshot();
  });

  test('3', () => {
    let result: any = findBestPasses.findBestPasses('Jean-Philippe,George', {
      static: true,
      linkIridium: false,
      linkGalileo: true,
      linkStarlink: true,
      obsmaxel2: 0,
      obsmaxrange2: 70,
      obsminrange2: 1,
      obsminel2: 1,
      obsmaxaz2: -5.48,
      obsminaz2: -100,
      alt: 5,
      beamwidth: 16,
      changeObjectInterval: 0,
      country: 'France',
      lat: -45,
      linkAehf: true,
      linkWgs: true,
      lon: 45,
      name: 'Anas',
      observerGd: { lat: 45, lon: 90, alt: 5 },
      obsmaxaz: 0,
      obsmaxel: 0,
      obsmaxrange: 520,
      obsminaz: 1,
      obsminel: -1.0,
      obsminrange: 100,
      shortName: 'http://example.com/showcalendar.html?token=CKF50YzIHxCTKMAg',
      staticNum: 10,
      sun: '1.0.0',
      type: SpaceObjectType.SpaceObjectType.MECHANICAL,
      url: 'http://base.com',
      volume: false,
      zoom: 'leo',
    });
    expect(result).toMatchSnapshot();
  });

  test('4', () => {
    let result: any = findBestPasses.findBestPasses(',Jean-Philippe', {
      static: true,
      linkIridium: false,
      linkGalileo: false,
      linkStarlink: true,
      obsmaxel2: 1,
      obsmaxrange2: 4,
      obsminrange2: 50,
      obsminel2: 380,
      obsmaxaz2: 100,
      obsminaz2: 1,
      alt: 3,
      beamwidth: 24,
      changeObjectInterval: -100,
      country: 'GB',
      lat: -90,
      linkAehf: false,
      linkWgs: false,
      lon: 90,
      name: 'Edmond',
      observerGd: { lat: -180, lon: 45, alt: 50 },
      obsmaxaz: -100,
      obsmaxel: 1,
      obsmaxrange: 90,
      obsminaz: 100,
      obsminel: -1.0,
      obsminrange: 100,
      shortName: 'https://twitter.com/path?abc',
      staticNum: 10,
      sun: 'v4.0.0-rc.4',
      type: SpaceObjectType.SpaceObjectType.LAUNCH_VEHICLE_MANUFACTURER,
      url: 'http://www.example.com/route/123?foo=bar',
      volume: true,
      zoom: 'leo',
    });
    expect(result).toMatchSnapshot();
  });
});
