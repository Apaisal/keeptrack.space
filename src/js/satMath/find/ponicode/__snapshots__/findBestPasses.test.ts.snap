// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findBestPasses.findBestPasses 0 1`] = `undefined`;

exports[`findBestPasses.findBestPasses 1 1`] = `undefined`;

exports[`findBestPasses.findBestPasses 2 1`] = `undefined`;

exports[`findBestPasses.findBestPasses 3 1`] = `undefined`;

exports[`findBestPasses.findBestPasses 4 1`] = `undefined`;
